# Backendå®Ÿè£…è¦ç´„ (Controller + Service)

## ğŸ“‹ ç›®æ¬¡
- [Backendå®Ÿè£…è¦ç´„ (Controller + Service)](#backendå®Ÿè£…è¦ç´„-controller-service) (L1-29)
  - [1. Controllerå±¤ã®åŸå‰‡](#1-controllerå±¤ã®åŸå‰‡) (L10-29)
    - [1.1 è²¬å‹™](#11-è²¬å‹™) (L12-24)
    - [1.2 å¿…é ˆãƒ‘ã‚¿ãƒ¼ãƒ³](#12-å¿…é ˆãƒ‘ã‚¿ãƒ¼ãƒ³) (L26-29)
- [app/controllers/concerns/kanban_api_concern.rb ã«å…±é€šå‡¦ç†ã‚’é›†ç´„](#appcontrollersconcernskanban_api_concernrb-ã«å…±é€šå‡¦ç†ã‚’é›†ç´„) (L30-62)
    - [1.3 ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹è¦ä»¶](#13-ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹è¦ä»¶) (L39-48)
    - [1.4 ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£å¿…é ˆé …ç›®](#14-ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£å¿…é ˆé …ç›®) (L50-62)
- [Strong Parameterså¿…é ˆ](#strong-parameterså¿…é ˆ) (L63-64)
- [ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆé…ä¸‹ã®ãƒªã‚½ãƒ¼ã‚¹æ¤œè¨¼å¿…é ˆ](#ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆé…ä¸‹ã®ãƒªã‚½ãƒ¼ã‚¹æ¤œè¨¼å¿…é ˆ) (L66-221)
    - [1.5 ãƒ¬ã‚¹ãƒãƒ³ã‚¹å½¢å¼çµ±ä¸€](#15-ãƒ¬ã‚¹ãƒãƒ³ã‚¹å½¢å¼çµ±ä¸€) (L70-80)
    - [1.6 å‚ç…§å®Ÿè£…](#16-å‚ç…§å®Ÿè£…) (L82-86)
  - [2. Serviceå±¤ã®åŸå‰‡](#2-serviceå±¤ã®åŸå‰‡) (L88-192)
    - [2.1 è²¬å‹™](#21-è²¬å‹™) (L90-101)
    - [2.2 å˜ä¸€è²¬ä»»ã®åŸå‰‡](#22-å˜ä¸€è²¬ä»»ã®åŸå‰‡) (L103-122)
    - [2.3 æˆ»ã‚Šå€¤çµ±ä¸€è¦ç´„](#23-æˆ»ã‚Šå€¤çµ±ä¸€è¦ç´„) (L124-136)
    - [2.4 ã‚«ã‚¹ã‚¿ãƒ ä¾‹å¤–å®šç¾©](#24-ã‚«ã‚¹ã‚¿ãƒ ä¾‹å¤–å®šç¾©) (L138-148)
    - [2.5 ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³åˆ¶å¾¡](#25-ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³åˆ¶å¾¡) (L150-163)
    - [2.6 N+1å¯¾ç­–ãƒ‘ã‚¿ãƒ¼ãƒ³](#26-n1å¯¾ç­–ãƒ‘ã‚¿ãƒ¼ãƒ³) (L165-172)
    - [2.7 å†ªç­‰æ€§ä¿è¨¼](#27-å†ªç­‰æ€§ä¿è¨¼) (L174-186)
    - [2.8 å‚ç…§å®Ÿè£…](#28-å‚ç…§å®Ÿè£…) (L188-192)
  - [3. å®Ÿè£…ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ](#3-å®Ÿè£…ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ) (L194-214)
    - [Controllerå®Ÿè£…æ™‚](#controllerå®Ÿè£…æ™‚) (L196-203)
    - [Serviceå®Ÿè£…æ™‚](#serviceå®Ÿè£…æ™‚) (L205-214)
  - [ğŸ”— é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ](#-é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ) (L216-221)
## 1. Controllerå±¤ã®åŸå‰‡

### 1.1 è²¬å‹™

**ã‚„ã‚‹ã¹ãã“ã¨**:
- HTTPè¦æ±‚å—ä»˜ãƒ»ãƒ¬ã‚¹ãƒãƒ³ã‚¹åˆ¶å¾¡
- æ¨©é™æ¤œè¨¼ (`before_action :authorize_kanban_access`)
- ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿æ¤œè¨¼ (Strong Parameters)
- Serviceå±¤å‘¼ã³å‡ºã—
- çµæœã®ã‚·ãƒªã‚¢ãƒ©ã‚¤ã‚¼ãƒ¼ã‚·ãƒ§ãƒ³ (JSONåŒ–)

**ã‚„ã£ã¦ã¯ã„ã‘ãªã„ã“ã¨**:
- âŒ ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯å®Ÿè£…
- âŒ ç›´æ¥DBæ“ä½œ (`Issue.find.update!`ãªã©)
- âŒ è¤‡é›‘ãªãƒ‡ãƒ¼ã‚¿å¤‰æ›

### 1.2 å¿…é ˆãƒ‘ã‚¿ãƒ¼ãƒ³

**Concernæ´»ç”¨**:
```ruby
# app/controllers/concerns/kanban_api_concern.rb ã«å…±é€šå‡¦ç†ã‚’é›†ç´„
include KanbanApiConcern
```

**ã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°éšå±¤**:
1. **å°‚ç”¨ä¾‹å¤–** - `KanbanService::InvalidTransitionError` â†’ 400 Bad Request
2. **æ±ç”¨ä¾‹å¤–** - `ActiveRecord::RecordNotFound` â†’ 404 Not Found
3. **äºˆæœŸå¤–** - `StandardError` â†’ 500 Internal Server Error

### 1.3 ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹è¦ä»¶

**N+1å¯¾ç­–å¿…é ˆ**:
- `includes(:tracker, :status, :assigned_to, :fixed_version)` ã§äº‹å‰èª­ã¿è¾¼ã¿
- Bulletãƒ„ãƒ¼ãƒ«ã§æ¤œå‡ºãƒ»ä¿®æ­£

**APIå¿œç­”æ™‚é–“åŸºæº–**:
- å¹³å‡: **200msä»¥å†…**
- 95ãƒ‘ãƒ¼ã‚»ãƒ³ã‚¿ã‚¤ãƒ«: **500msä»¥å†…**
- ã‚¯ã‚¨ãƒªæ•°: **3ã‚¯ã‚¨ãƒªä»¥ä¸‹**

### 1.4 ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£å¿…é ˆé …ç›®

**æ¨©é™ãƒã‚§ãƒƒã‚¯**:
```ruby
before_action :authorize_kanban_access

def authorize_kanban_access
  deny_access unless User.current.allowed_to?(:view_kanban, @project)
end
```

**ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿æ¤œè¨¼**:
```ruby
# Strong Parameterså¿…é ˆ
params.permit(:version_id, :assignee_id, tracker_ids: [])

# ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆé…ä¸‹ã®ãƒªã‚½ãƒ¼ã‚¹æ¤œè¨¼å¿…é ˆ
@project.trackers.pluck(:id).include?(params[:tracker_id])
```

### 1.5 ãƒ¬ã‚¹ãƒãƒ³ã‚¹å½¢å¼çµ±ä¸€

**æˆåŠŸ**:
```ruby
{ success: true, data: result, meta: { total_count: count } }
```

**ã‚¨ãƒ©ãƒ¼**:
```ruby
{ success: false, error: message, error_code: 'VALIDATION_ERROR' }
```

### 1.6 å‚ç…§å®Ÿè£…

**å®Ÿè£…æ¸ˆã¿Controller**: `app/controllers/kanban/`é…ä¸‹ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å‚ç…§ã—ã¦ãã ã•ã„ã€‚

---

## 2. Serviceå±¤ã®åŸå‰‡

### 2.1 è²¬å‹™

**ã‚„ã‚‹ã¹ãã“ã¨**:
- ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯å®Ÿè£…
- ãƒ‡ãƒ¼ã‚¿å¤‰æ›ãƒ»åŠ å·¥
- è¤‡é›‘ãªè¨ˆç®—ãƒ»åˆ¤å®š
- éåŒæœŸJobå‘¼ã³å‡ºã—

**ã‚„ã£ã¦ã¯ã„ã‘ãªã„ã“ã¨**:
- âŒ HTTPè¦æ±‚å‡¦ç† (Controllerè²¬å‹™)
- âŒ Viewæç”»å‡¦ç†
- âŒ ç›´æ¥æ¨©é™ãƒã‚§ãƒƒã‚¯ (Controllerè²¬å‹™)

### 2.2 å˜ä¸€è²¬ä»»ã®åŸå‰‡

**Good**: 1ã¤ã®Serviceã¯1ã¤ã®è²¬å‹™ã®ã¿
```ruby
class Kanban::VersionPropagationService
  # ãƒãƒ¼ã‚¸ãƒ§ãƒ³ä¼æ’­ã®ã¿ã«ç‰¹åŒ–
end

class Kanban::TestGenerationService
  # Testè‡ªå‹•ç”Ÿæˆã®ã¿ã«ç‰¹åŒ–
end
```

**Bad**: è¤‡æ•°è²¬å‹™æ··åœ¨
```ruby
class BadKanbanService
  def do_everything  # ãƒãƒ¼ã‚¸ãƒ§ãƒ³ãƒ»çŠ¶æ…‹ãƒ»æ¤œè¨¼ã™ã¹ã¦ âŒ
  end
end
```

### 2.3 æˆ»ã‚Šå€¤çµ±ä¸€è¦ç´„

**å¿…é ˆ**: å…¨Serviceã¯ä»¥ä¸‹ã®å½¢å¼ã§æˆ»ã‚Šå€¤ã‚’çµ±ä¸€

**æˆåŠŸæ™‚**:
```ruby
{ success: true, data: result_data, meta: additional_info }
```

**å¤±æ•—æ™‚**:
```ruby
{ success: false, error: error_message, error_code: 'CODE', details: {...} }
```

### 2.4 ã‚«ã‚¹ã‚¿ãƒ ä¾‹å¤–å®šç¾©

**æ¥­å‹™å›ºæœ‰ã‚¨ãƒ©ãƒ¼ã¯å°‚ç”¨ä¾‹å¤–ã‚’å®šç¾©**:
```ruby
module Kanban
  class StateTransitionService
    class InvalidTransitionError < StandardError; end
    class TransitionBlockedError < StandardError; end
  end
end
```

### 2.5 ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³åˆ¶å¾¡

**è¤‡æ•°æ›´æ–°ã¯å¿…ãšãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³å†…ã§å®Ÿè¡Œ**:
```ruby
def execute
  ActiveRecord::Base.transaction do
    step1_result = execute_step1
    step2_result = execute_step2(step1_result)
    { success: true, data: step2_result }
  end
rescue CustomError => e
  { success: false, error: e.message, error_code: 'CUSTOM_ERROR' }
end
```

### 2.6 N+1å¯¾ç­–ãƒ‘ã‚¿ãƒ¼ãƒ³

**å¿…é ˆ**:
- `includes` - é–¢é€£ãƒ‡ãƒ¼ã‚¿äº‹å‰èª­ã¿è¾¼ã¿
- `pluck` - å¿…è¦ã‚«ãƒ©ãƒ ã®ã¿å–å¾—
- `find_each(batch_size: 100)` - å¤§é‡ãƒ‡ãƒ¼ã‚¿å‡¦ç†

**å®Ÿè£…ä¾‹ã¯å®Ÿè£…æ¸ˆã¿Serviceã‚’å‚ç…§**ã—ã¦ãã ã•ã„ã€‚

### 2.7 å†ªç­‰æ€§ä¿è¨¼

**éåŒæœŸJobãƒ»APIå‘¼ã³å‡ºã—ã¯å†ªç­‰æ€§ã‚’ç¢ºä¿**:
```ruby
def idempotent_operation(user_story)
  # æ—¢å­˜ãƒ¬ã‚³ãƒ¼ãƒ‰ç¢ºèª
  existing = find_existing_test(user_story)
  return existing if existing && !options[:force_recreate]

  # ä½œæˆå‡¦ç†
  create_new_test(user_story)
end
```

### 2.8 å‚ç…§å®Ÿè£…

**å®Ÿè£…æ¸ˆã¿Service**: `app/services/kanban/`é…ä¸‹ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å‚ç…§ã—ã¦ãã ã•ã„ã€‚

---

## 3. å®Ÿè£…ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

### Controllerå®Ÿè£…æ™‚

- [ ] `KanbanApiConcern` ã‚’includeã—ã¦ã„ã‚‹
- [ ] `before_action :authorize_kanban_access` ã‚’è¨­å®š
- [ ] Strong Parametersã§ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿æ¤œè¨¼
- [ ] Serviceå‘¼ã³å‡ºã—çµæœã‚’çµ±ä¸€å½¢å¼ã§ãƒ¬ã‚¹ãƒãƒ³ã‚¹
- [ ] ã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°3éšå±¤ã‚’å®Ÿè£…
- [ ] N+1å•é¡ŒãŒãªã„ã“ã¨ã‚’Bulletã§ç¢ºèª

### Serviceå®Ÿè£…æ™‚

- [ ] å˜ä¸€è²¬ä»»ã‚’å®ˆã£ã¦ã„ã‚‹ (1 Service = 1 è²¬å‹™)
- [ ] æˆ»ã‚Šå€¤ãŒçµ±ä¸€å½¢å¼ (`{success: true/false, ...}`)
- [ ] ã‚«ã‚¹ã‚¿ãƒ ä¾‹å¤–ã‚’å®šç¾©ã—ã¦ã„ã‚‹
- [ ] ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³å†…ã§è¤‡æ•°æ›´æ–°ã‚’å®Ÿè¡Œ
- [ ] N+1å¯¾ç­– (includes/pluck/find_each) ã‚’å®Ÿæ–½
- [ ] å†ªç­‰æ€§ã‚’ç¢ºä¿ (éåŒæœŸJobãƒ»APIå‘¼ã³å‡ºã—)

---

## ğŸ”— é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- **æŠ€è¡“ã‚¢ãƒ¼ã‚­ãƒ†ã‚¯ãƒãƒ£**: @vibes/rules/technical_architecture_quickstart.md
- **ãƒ†ã‚¹ãƒˆæˆ¦ç•¥**: @vibes/rules/testing_strategy.md
- **å®Ÿè£…æ¸ˆã¿Controller**: `app/controllers/kanban/`
- **å®Ÿè£…æ¸ˆã¿Service**: `app/services/kanban/`
