# ğŸ¯ Release Kanban ãƒ†ã‚¹ãƒˆæˆ¦ç•¥

## ğŸ“‹ æ¦‚è¦

Release Kanbanãƒ—ãƒ©ã‚°ã‚¤ãƒ³ã®7ã¤ã®ã‚³ã‚¢ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã«å¯¾ã™ã‚‹åŒ…æ‹¬çš„ãƒ†ã‚¹ãƒˆæˆ¦ç•¥ã¨ãƒ†ã‚¹ãƒˆãƒ”ãƒ©ãƒŸãƒƒãƒ‰å®Ÿè£…è¦ç´„

### ğŸ¯ ãƒ†ã‚¹ãƒˆå¯¾è±¡ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ

| ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ | é‡è¦åº¦ | èª¬æ˜ |
|---|---|---|
| ğŸ“Š TrackerHierarchy | ğŸ”´ Critical | Epicâ†’Featureâ†’UserStoryâ†’Task/Test ã®4æ®µéšéšå±¤åˆ¶ç´„ |
| ğŸ”„ VersionManagement | ğŸŸ¡ High | UserStoryã‹ã‚‰å­è¦ç´ ã¸ã®è‡ªå‹•ãƒãƒ¼ã‚¸ãƒ§ãƒ³ä¼æ’­ãƒ­ã‚¸ãƒƒã‚¯ |
| ğŸ¤– AutoGeneration | ğŸŸ¡ High | UserStoryä½œæˆæ™‚ã®Testè‡ªå‹•ç”Ÿæˆ + blocksé–¢ä¿‚ä½œæˆ |
| ğŸš¦ StateTransition | ğŸŸ¡ High | ã‚«ãƒ³ãƒãƒ³ã‚«ãƒ©ãƒ ç§»å‹•æ™‚ã®çŠ¶æ…‹é·ç§»åˆ¶å¾¡ |
| ğŸ›¡ï¸ ValidationGuard | ğŸ”´ Critical | 3å±¤ã‚¬ãƒ¼ãƒ‰æ¤œè¨¼ï¼ˆTaskå®Œäº†ãƒ»Teståˆæ ¼ãƒ»é‡å¤§Bugè§£æ±ºï¼‰ |
| ğŸ¨ KanbanUI | ğŸŸ¢ Medium | ãƒ‰ãƒ©ãƒƒã‚°&ãƒ‰ãƒ­ãƒƒãƒ—ã€Epic Swimlaneè¡¨ç¤º |
| ğŸ”Œ APIIntegration | ğŸ”´ Critical | React-Railsé–“ãƒ‡ãƒ¼ã‚¿äº¤æ›ã®æ­£ç¢ºæ€§ |

## ğŸ—ï¸ ãƒ†ã‚¹ãƒˆãƒ”ãƒ©ãƒŸãƒƒãƒ‰æ§‹é€ 

```
       /\
      /  \     Phase 4: System/E2E Tests
     /____\    - ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚¸ãƒ£ãƒ¼ãƒ‹ãƒ¼
    /      \   - ãƒ–ãƒ©ã‚¦ã‚¶çµ±åˆãƒ†ã‚¹ãƒˆ
   /        \  - ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ãƒ†ã‚¹ãƒˆ
  /__________\
 /            \ Phase 3: Integration Tests
/              \- APIçµ±åˆãƒ†ã‚¹ãƒˆ
\              /- ã‚µãƒ¼ãƒ“ã‚¹é–“é€£æºãƒ†ã‚¹ãƒˆ
 \____________/ - ã‚³ãƒ³ãƒˆãƒ­ãƒ¼ãƒ©ãƒ¼ãƒ†ã‚¹ãƒˆ
/              \
\              / Phase 2: Service Tests
 \____________/  - ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯ãƒ†ã‚¹ãƒˆ
/              \ - çŠ¶æ…‹é·ç§»ãƒ†ã‚¹ãƒˆ
\              / - æ¤œè¨¼ã‚¬ãƒ¼ãƒ‰ãƒ†ã‚¹ãƒˆ
 \____________/
/              \
\    Phase 1   / Unit Tests
 \____________/  - ãƒ¢ãƒ‡ãƒ«å˜ä½“ãƒ†ã‚¹ãƒˆ
                 - ãƒãƒªãƒ‡ãƒ¼ã‚·ãƒ§ãƒ³ãƒ†ã‚¹ãƒˆ
                 - ãƒ˜ãƒ«ãƒ‘ãƒ¼ãƒ†ã‚¹ãƒˆ
```

## ğŸ“Š Phase 1: å˜ä½“ãƒ†ã‚¹ãƒˆï¼ˆUnit Testsï¼‰

### ç›®çš„
- å€‹åˆ¥ã‚¯ãƒ©ã‚¹ãƒ»ãƒ¡ã‚½ãƒƒãƒ‰ã®å‹•ä½œä¿è¨¼
- ãƒ“ã‚¸ãƒã‚¹ãƒ«ãƒ¼ãƒ«ã®å³å¯†ãªæ¤œè¨¼
- é«˜é€Ÿãƒ•ã‚£ãƒ¼ãƒ‰ãƒãƒƒã‚¯ãƒ«ãƒ¼ãƒ—ã®æ§‹ç¯‰

### å¯¾è±¡
- `app/models/kanban/tracker_hierarchy.rb`
- `app/services/kanban/*_service.rb`
- `lib/kanban/helpers/*`

### ãƒ†ã‚¹ãƒˆè¦³ç‚¹
```ruby
# æ­£å¸¸ç³»
- æœŸå¾…ã•ã‚Œã‚‹å…¥åŠ›ã«å¯¾ã™ã‚‹æ­£ã—ã„å‡ºåŠ›
- ãƒ“ã‚¸ãƒã‚¹ãƒ«ãƒ¼ãƒ«ã®éµå®ˆ

# å¢ƒç•Œå€¤
- æœ€å¤§ãƒ»æœ€å°ãƒ»ç©ºå€¤ã§ã®å‹•ä½œ
- edge caseã®å‡¦ç†

# ç•°å¸¸ç³»
- ä¸æ­£å…¥åŠ›ã®æ‹’å¦
- é©åˆ‡ãªã‚¨ãƒ©ãƒ¼ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸

# ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹
- å¤§é‡ãƒ‡ãƒ¼ã‚¿ã§ã®æ€§èƒ½ç¶­æŒ
- ãƒ¡ãƒ¢ãƒªãƒªãƒ¼ã‚¯é˜²æ­¢
```

### å®Ÿè£…ä¾‹
```ruby
RSpec.describe Kanban::TrackerHierarchy do
  describe '.valid_parent?' do
    it 'æ­£å¸¸ãªè¦ªå­é–¢ä¿‚ã‚’è¨±å¯ã™ã‚‹' do
      expect(described_class.valid_parent?(task_tracker, user_story_tracker)).to be true
    end

    it 'ä¸æ­£ãªé–¢ä¿‚ã‚’æ‹’å¦ã™ã‚‹' do
      expect(described_class.valid_parent?(task_tracker, feature_tracker)).to be false
    end

    it 'nilå®‰å…¨æ€§ã‚’ä¿è¨¼ã™ã‚‹' do
      expect(described_class.valid_parent?(nil, user_story_tracker)).to be false
    end
  end
end
```

## ğŸ¤– Phase 2: ã‚µãƒ¼ãƒ“ã‚¹ãƒ†ã‚¹ãƒˆï¼ˆService Testsï¼‰

### ç›®çš„
- ãƒ“ã‚¸ãƒã‚¹ãƒ­ã‚¸ãƒƒã‚¯ã®çµ±åˆå‹•ä½œç¢ºèª
- ã‚µãƒ¼ãƒ“ã‚¹é–“ã®é€£æºæ¤œè¨¼
- ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³åˆ¶å¾¡ã®ç¢ºèª

### å¯¾è±¡
- `app/services/kanban/test_generation_service.rb`
- `app/services/kanban/version_propagation_service.rb`
- `app/services/kanban/state_transition_service.rb`
- `app/services/kanban/validation_guard_service.rb`

### ãƒ†ã‚¹ãƒˆè¦³ç‚¹
```ruby
# ã‚µãƒ¼ãƒ“ã‚¹é€£æº
- è¤‡æ•°ã‚µãƒ¼ãƒ“ã‚¹ã®å”èª¿å‹•ä½œ
- ãƒ‡ãƒ¼ã‚¿æ•´åˆæ€§ã®ç¶­æŒ

# ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³
- æˆåŠŸæ™‚ã®ã‚³ãƒŸãƒƒãƒˆ
- å¤±æ•—æ™‚ã®ãƒ­ãƒ¼ãƒ«ãƒãƒƒã‚¯

# çŠ¶æ…‹å¤‰åŒ–
- å‰çŠ¶æ…‹ã¨å¾ŒçŠ¶æ…‹ã®ç¢ºèª
- å‰¯ä½œç”¨ã®æ¤œè¨¼

# ã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°
- é©åˆ‡ãªã‚¨ãƒ©ãƒ¼æƒ…å ±è¿”å´
- ãƒ­ã‚°å‡ºåŠ›ã®ç¢ºèª
```

### å®Ÿè£…ä¾‹
```ruby
RSpec.describe Kanban::TestGenerationService do
  describe '.generate_test_for_user_story' do
    it 'Testã‚’ç”Ÿæˆã—blocksé–¢ä¿‚ã‚’ä½œæˆã™ã‚‹' do
      result = described_class.generate_test_for_user_story(user_story)

      expect(result[:test_issue]).to be_a(Issue)
      expect(result[:relation_created]).to be true

      # blocksé–¢ä¿‚ã®ç¢ºèª
      relation = IssueRelation.find_by(
        issue_from: result[:test_issue],
        issue_to: user_story,
        relation_type: 'blocks'
      )
      expect(relation).to be_present
    end

    it 'ã‚¨ãƒ©ãƒ¼æ™‚ã¯ãƒ­ãƒ¼ãƒ«ãƒãƒƒã‚¯ã•ã‚Œã‚‹' do
      allow(IssueRelation).to receive(:create!).and_raise(ActiveRecord::RecordInvalid)

      initial_count = Issue.count
      result = described_class.generate_test_for_user_story(user_story)

      expect(result[:error]).to be_present
      expect(Issue.count).to eq(initial_count)
    end
  end
end
```

## ğŸ”Œ Phase 3: çµ±åˆãƒ†ã‚¹ãƒˆï¼ˆIntegration Testsï¼‰

### ç›®çš„
- APIå±¤ã‹ã‚‰ã‚µãƒ¼ãƒ“ã‚¹å±¤ã¾ã§ã®çµ±åˆå‹•ä½œç¢ºèª
- React-Railsé–“ã®ãƒ‡ãƒ¼ã‚¿äº¤æ›æ¤œè¨¼
- æ¨©é™ãƒ»èªè¨¼ã‚·ã‚¹ãƒ†ãƒ ã®ç¢ºèª

### å¯¾è±¡
- `app/controllers/kanban/api_controller.rb`
- `app/controllers/kanban_controller.rb`
- APIã‚¨ãƒ³ãƒ‰ãƒã‚¤ãƒ³ãƒˆå…¨ä½“
- ã‚µãƒ¼ãƒ“ã‚¹é–“ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼

### ãƒ†ã‚¹ãƒˆè¦³ç‚¹
```ruby
# APIçµ±åˆ
- HTTPãƒªã‚¯ã‚¨ã‚¹ãƒˆãƒ»ãƒ¬ã‚¹ãƒãƒ³ã‚¹
- JSONå½¢å¼ã®æ­£ç¢ºæ€§
- ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ã‚³ãƒ¼ãƒ‰ã®é©åˆ‡æ€§

# æ¨©é™åˆ¶å¾¡
- èªè¨¼ãƒ»èªå¯ã®å‹•ä½œ
- ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã‚¢ã‚¯ã‚»ã‚¹åˆ¶å¾¡
- ãƒ­ãƒ¼ãƒ«ãƒ™ãƒ¼ã‚¹æ¨©é™

# ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼çµ±åˆ
- è¤‡æ•°ã‚µãƒ¼ãƒ“ã‚¹ã®é€£ç¶šå®Ÿè¡Œ
- ãƒ‡ãƒ¼ã‚¿æµã‚Œã®ç¢ºèª
- ã‚¨ãƒ©ãƒ¼ä¼æ’­ã®æ¤œè¨¼
```

### å®Ÿè£…ä¾‹
```ruby
RSpec.describe Kanban::ApiController, type: :request do
  describe 'POST /kanban/api/transition_issue' do
    it 'UserStoryã®çŠ¶æ…‹é·ç§»ãŒæˆåŠŸã™ã‚‹' do
      post "/kanban/api/transition_issue", params: {
        project_id: project.id,
        issue_id: user_story.id,
        target_column: 'ready'
      }

      expect(response).to have_http_status(:success)
      json = JSON.parse(response.body)
      expect(json['success']).to be true
      expect(json['issue']['status']).to eq('Ready')

      user_story.reload
      expect(user_story.status.name).to eq('Ready')
    end

    it 'ãƒ–ãƒ­ãƒƒã‚¯æ¡ä»¶é•åæ™‚ã¯ã‚¨ãƒ©ãƒ¼ã‚’è¿”ã™' do
      incomplete_task = create(:task, parent: user_story, status: 'New')

      post "/kanban/api/transition_issue", params: {
        project_id: project.id,
        issue_id: user_story.id,
        target_column: 'done'
      }

      expect(response).to have_http_status(:unprocessable_entity)
      json = JSON.parse(response.body)
      expect(json['error']).to include('æœªå®Œäº†ã®Task')
    end
  end
end
```

## ğŸ¨ Phase 4: ã‚·ã‚¹ãƒ†ãƒ ãƒ†ã‚¹ãƒˆï¼ˆSystem/E2E Testsï¼‰

### ç›®çš„
- ãƒ¦ãƒ¼ã‚¶ãƒ¼è¦–ç‚¹ã§ã®ã‚¨ãƒ³ãƒ‰ãƒ„ãƒ¼ã‚¨ãƒ³ãƒ‰å‹•ä½œç¢ºèª
- ãƒ–ãƒ©ã‚¦ã‚¶ã§ã®å®Ÿéš›ã®æ“ä½œã‚·ãƒŸãƒ¥ãƒ¬ãƒ¼ã‚·ãƒ§ãƒ³
- ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ãƒ»ã‚¢ã‚¯ã‚»ã‚·ãƒ“ãƒªãƒ†ã‚£ã®æ¤œè¨¼

### å¯¾è±¡
- ã‚«ãƒ³ãƒãƒ³ãƒšãƒ¼ã‚¸å…¨ä½“
- ãƒ‰ãƒ©ãƒƒã‚°&ãƒ‰ãƒ­ãƒƒãƒ—æ“ä½œ
- Epic Swimlaneè¡¨ç¤º
- ãƒ•ã‚£ãƒ«ã‚¿ãƒªãƒ³ã‚°æ©Ÿèƒ½

### ãƒ†ã‚¹ãƒˆè¦³ç‚¹
```ruby
# ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚¸ãƒ£ãƒ¼ãƒ‹ãƒ¼
- å…¸å‹çš„ãªä½œæ¥­ãƒ•ãƒ­ãƒ¼ã®å®Œå…¨å®Ÿè¡Œ
- è¤‡æ•°ç”»é¢ã«ã‚ãŸã‚‹æ“ä½œ

# ãƒ–ãƒ©ã‚¦ã‚¶çµ±åˆ
- JavaScriptå‹•ä½œç¢ºèª
- DOMæ“ä½œã®æ¤œè¨¼

# ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹
- ãƒšãƒ¼ã‚¸èª­ã¿è¾¼ã¿æ™‚é–“
- å¤§é‡ãƒ‡ãƒ¼ã‚¿ã§ã®å¿œç­”æ€§

# ã‚¢ã‚¯ã‚»ã‚·ãƒ“ãƒªãƒ†ã‚£
- ã‚­ãƒ¼ãƒœãƒ¼ãƒ‰ãƒŠãƒ“ã‚²ãƒ¼ã‚·ãƒ§ãƒ³
- ã‚¹ã‚¯ãƒªãƒ¼ãƒ³ãƒªãƒ¼ãƒ€ãƒ¼å¯¾å¿œ
```

### å®Ÿè£…ä¾‹
```ruby
RSpec.describe 'Release Kanban System', type: :system, js: true do
  scenario 'ãƒ¦ãƒ¼ã‚¶ãƒ¼ãŒã‚«ãƒ¼ãƒ‰ã‚’ãƒ‰ãƒ©ãƒƒã‚°&ãƒ‰ãƒ­ãƒƒãƒ—ã§ç§»å‹•ã™ã‚‹' do
    visit "/projects/#{project.identifier}/kanban"

    # Epic Swimlaneã®è¡¨ç¤ºç¢ºèª
    expect(page).to have_selector('.epic-swimlane')

    # ã‚«ãƒ¼ãƒ‰ã®ãƒ‰ãƒ©ãƒƒã‚°&ãƒ‰ãƒ­ãƒƒãƒ—
    user_story_card = find('.issue-card[data-tracker="UserStory"]')
    in_progress_column = find('.kanban-column[data-column-id="in_progress"]')

    user_story_card.drag_to(in_progress_column)

    # çŠ¶æ…‹æ›´æ–°ã®ç¢ºèª
    expect(page).to have_content('In Progress')

    # ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹æ›´æ–°ã®ç¢ºèª
    user_story.reload
    expect(user_story.status.name).to eq('In Progress')
  end

  scenario 'Testè‡ªå‹•ç”Ÿæˆæ©Ÿèƒ½ãŒæ­£å¸¸å‹•ä½œã™ã‚‹' do
    visit "/projects/#{project.identifier}/kanban"

    within('.issue-card[data-tracker="UserStory"]') do
      click_button 'Testä½œæˆ'
    end

    # Testã‚«ãƒ¼ãƒ‰ã®è¡¨ç¤ºç¢ºèª
    expect(page).to have_selector('.issue-card[data-tracker="Test"]')

    # ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ä½œæˆç¢ºèª
    test_issue = Issue.joins(:tracker).find_by(
      trackers: { name: 'Test' },
      parent: user_story
    )
    expect(test_issue).to be_present
  end
end
```

## ğŸ“ˆ å“è³ªåŸºæº–ãƒ»æˆåŠŸæŒ‡æ¨™

### ã‚«ãƒãƒ¬ãƒƒã‚¸ç›®æ¨™
```yaml
å…¨ä½“ã‚«ãƒãƒ¬ãƒƒã‚¸: 80%ä»¥ä¸Š
Criticalè¦ç´ : 95%ä»¥ä¸Š
å„ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ: 85%ä»¥ä¸Š

è©³ç´°:
  - models/: 90%ä»¥ä¸Š
  - services/: 85%ä»¥ä¸Š
  - controllers/: 80%ä»¥ä¸Š
  - javascripts/: 75%ä»¥ä¸Š
```

### ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åŸºæº–
```yaml
å®Ÿè¡Œæ™‚é–“:
  - å…¨ãƒ†ã‚¹ãƒˆ: 5åˆ†ä»¥å†…
  - Phase 1+2: 2åˆ†ä»¥å†…
  - Phase 3: 1åˆ†ä»¥å†…
  - Phase 4: 2åˆ†ä»¥å†…

APIå¿œç­”:
  - é€šå¸¸API: 200msä»¥å†…
  - è¤‡é›‘å‡¦ç†: 500msä»¥å†…
  - å¤§é‡ãƒ‡ãƒ¼ã‚¿: 1ç§’ä»¥å†…
```

### å®‰å®šæ€§åŸºæº–
```yaml
CIæˆåŠŸç‡: 95%ä»¥ä¸Š
ãƒ•ãƒ¬ãƒ¼ã‚­ãƒ¼ãƒ†ã‚¹ãƒˆ: 0å€‹
å†ç¾æ€§: 100%ï¼ˆåŒæ¡ä»¶ã§å¿…ãšåŒçµæœï¼‰
```

## ğŸ› ï¸ ãƒ†ã‚¹ãƒˆç’°å¢ƒãƒ»ãƒ„ãƒ¼ãƒ«

### å¿…è¦ãªGem
```ruby
group :test do
  gem 'rspec-rails', '~> 5.0'
  gem 'factory_bot_rails', '~> 6.2'
  gem 'capybara', '~> 3.0'
  gem 'selenium-webdriver', '~> 4.0'
  gem 'shoulda-matchers', '~> 5.3'
  gem 'rspec-benchmark', '~> 0.6'
  gem 'timecop', '~> 0.9'
  gem 'database_cleaner', '~> 2.0'
  gem 'simplecov', '~> 0.21'
end
```

### ãƒ–ãƒ©ã‚¦ã‚¶ç’°å¢ƒ
```bash
# ChromeDriverï¼ˆå¿…é ˆï¼‰
sudo apt-get install chromium-chromedriver

# ã¾ãŸã¯
npm install -g chromedriver
```

### Capybaraè¨­å®š
```ruby
# spec/rails_helper.rb
Capybara.default_driver = :selenium_chrome_headless
Capybara.javascript_driver = :selenium_chrome_headless
Capybara.default_max_wait_time = 5
```

## ğŸš€ ãƒ†ã‚¹ãƒˆå®Ÿè¡Œæˆ¦ç•¥

### é–‹ç™ºãƒ•ãƒ­ãƒ¼çµ±åˆ
```bash
# 1. ãƒ­ãƒ¼ã‚«ãƒ«é–‹ç™ºæ™‚
./bin/test_runner.sh quick    # Phase 1+2ã®ã¿é«˜é€Ÿå®Ÿè¡Œ

# 2. ãƒ•ã‚£ãƒ¼ãƒãƒ£ãƒ¼å®Œæˆæ™‚
./bin/test_runner.sh unit     # è©³ç´°ãªå˜ä½“ãƒ»ã‚µãƒ¼ãƒ“ã‚¹ãƒ†ã‚¹ãƒˆ

# 3. PRä½œæˆå‰
./bin/test_runner.sh full     # å…¨Phaseå®Ÿè¡Œ

# 4. ãƒªãƒªãƒ¼ã‚¹å‰
./bin/test_runner.sh full && run_performance_tests
```

### CI/CDçµ±åˆ
```yaml
# .github/workflows/test.yml
strategy:
  matrix:
    phase: [phase1, phase2, phase3, phase4]
parallel: true
timeout: 10åˆ†
```

### æ®µéšçš„ãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
```bash
# ç·Šæ€¥æ™‚ã®æ®µéšå®Ÿè¡Œ
./bin/test_runner.sh phase1   # Criticalè¦ç´ ã®ã¿
./bin/test_runner.sh phase2   # + ã‚µãƒ¼ãƒ“ã‚¹å±¤
./bin/test_runner.sh phase3   # + APIçµ±åˆ
./bin/test_runner.sh phase4   # + E2Eï¼ˆãƒ•ãƒ«å®Ÿè¡Œï¼‰
```

## ğŸ”§ ãƒ‡ãƒãƒƒã‚°ãƒ»ãƒˆãƒ©ãƒ–ãƒ«ã‚·ãƒ¥ãƒ¼ãƒ†ã‚£ãƒ³ã‚°

### ã‚ˆãã‚ã‚‹å•é¡Œã¨è§£æ±ºæ³•

#### 1. System/E2Eãƒ†ã‚¹ãƒˆã®ä¸å®‰å®šæ€§
```ruby
# è§£æ±ºç­–: é©åˆ‡ãªå¾…æ©Ÿå‡¦ç†
expect(page).to have_selector('.issue-card', wait: 10)

# å‹•çš„è¦ç´ ã®ç¢ºå®Ÿãªå¾…æ©Ÿ
wait_for { page.has_content?('Expected Text') }
```

#### 2. ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹çŠ¶æ…‹ã®ä¸æ•´åˆ
```ruby
# è§£æ±ºç­–: ãƒˆãƒ©ãƒ³ã‚¶ã‚¯ã‚·ãƒ§ãƒ³åˆ†é›¢
config.use_transactional_fixtures = true

# ã¾ãŸã¯æ˜ç¤ºçš„ã‚¯ãƒªãƒ¼ãƒ³ã‚¢ãƒƒãƒ—
after(:each) do
  DatabaseCleaner.clean
end
```

#### 3. éåŒæœŸå‡¦ç†ã®ãƒ†ã‚¹ãƒˆ
```ruby
# è§£æ±ºç­–: Timecopæ´»ç”¨
Timecop.freeze(Time.current) do
  # ãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
end

# éåŒæœŸã‚¸ãƒ§ãƒ–ã®ãƒ†ã‚¹ãƒˆ
expect(TestGenerationJob).to have_been_enqueued
```

### ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹æ”¹å–„
```ruby
# 1. ãƒ†ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ã®æœ€å°åŒ–
let!(:minimal_user_story) { build_minimal(:user_story) }

# 2. ä¸¦åˆ—å®Ÿè¡Œã®æ´»ç”¨
RSpec.configure do |config|
  config.default_formatter = 'ParallelTests::RSpec::RuntimeLogger'
end

# 3. é‡ã„ã‚»ãƒƒãƒˆã‚¢ãƒƒãƒ—ã®å…±æœ‰
before(:all) do
  @shared_project = create(:project_with_trackers)
end
```

## ğŸ“ ãƒ†ã‚¹ãƒˆè¦ç´„éµå®ˆãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

### PRä½œæˆå‰ãƒã‚§ãƒƒã‚¯
- [ ] æ–°æ©Ÿèƒ½ã«å¯¾å¿œã™ã‚‹ãƒ†ã‚¹ãƒˆã‚’å…¨Phaseä½œæˆ
- [ ] ã‚«ãƒãƒ¬ãƒƒã‚¸åŸºæº–ã‚’æº€ãŸã—ã¦ã„ã‚‹
- [ ] CI/CDãŒå…¨ã¦ç·‘è‰²ã§é€šé
- [ ] ãƒ•ãƒ¬ãƒ¼ã‚­ãƒ¼ãƒ†ã‚¹ãƒˆãŒ0å€‹
- [ ] ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åŸºæº–ã‚’æº€ãŸã—ã¦ã„ã‚‹

### ãƒªãƒªãƒ¼ã‚¹å‰ãƒã‚§ãƒƒã‚¯
- [ ] æœ¬ç•ªé¡ä¼¼ç’°å¢ƒã§ã®E2Eãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
- [ ] è² è·ãƒ†ã‚¹ãƒˆã®å®Ÿæ–½
- [ ] ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£ãƒ†ã‚¹ãƒˆã®ç¢ºèª
- [ ] ãƒ–ãƒ©ã‚¦ã‚¶äº’æ›æ€§ãƒ†ã‚¹ãƒˆã®å®Œäº†
- [ ] ã‚¢ã‚¯ã‚»ã‚·ãƒ“ãƒªãƒ†ã‚£ãƒ†ã‚¹ãƒˆã®é€šé

### ã‚³ãƒ¼ãƒ‰ãƒ¬ãƒ“ãƒ¥ãƒ¼è¦³ç‚¹
- [ ] ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹ã®ç¶²ç¾…æ€§
- [ ] ãƒ†ã‚¹ãƒˆã®å¯èª­æ€§ãƒ»ä¿å®ˆæ€§
- [ ] é©åˆ‡ãªãƒ†ã‚¹ãƒˆç²’åº¦ã®é¸æŠ
- [ ] ãƒ¢ãƒƒã‚¯ã®é©åˆ‡ãªä½¿ç”¨
- [ ] ãƒ†ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ã®é©åˆ‡æ€§

---

**ğŸ¯ æœ€çµ‚ç›®æ¨™**: ã“ã®ãƒ†ã‚¹ãƒˆæˆ¦ç•¥ã«ã‚ˆã‚Šã€Release Kanbanã®å“è³ªå‘ä¸Šãƒ»é–‹ç™ºåŠ¹ç‡30%æ”¹å–„ãƒ»ãƒã‚°æ··å…¥ãƒªã‚¹ã‚¯80%å‰Šæ¸›ã‚’å®Ÿç¾ã—ã€å®‰å¿ƒã—ã¦ãƒªãƒªãƒ¼ã‚¹ã§ãã‚‹ä½“åˆ¶ã‚’æ§‹ç¯‰ã™ã‚‹